include/log.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <ostream>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <variant>
+
+// TODO: Make this configurable through CMake?
+#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE
+#include <spdlog/spdlog.h>
+
+#include <spdlog/fmt/ostr.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+
+#include "print_utils.h"
+
+#define CELERITY_SET_LOG_CTX(ctx) CELERITY_DETAIL_SET_LOG_CTX(ctx)
+
+#define CELERITY_TRACE(...) SPDLOG_TRACE(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_DEBUG(...) SPDLOG_DEBUG(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_INFO(...) SPDLOG_INFO(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_WARN(...) SPDLOG_WARN(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_ERROR(...) SPDLOG_ERROR(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_CRITICAL(...) SPDLOG_CRITICAL(*celerity::detail::active_log_ctx + __VA_ARGS__)
+
+namespace celerity {
+namespace detail {
+
+	using log_map = std::unordered_map<std::string, std::variant<std::string_view, size_t>>;
+	using log_level = spdlog::level::level_enum;
+
+	struct log_context {
+		std::string value;
+		log_context() = default;
+		explicit log_context(const log_map& values) { value = fmt::format("[{}] ", values); }
+	};
+
+	inline const std::string null_log_ctx;
+	inline thread_local const std::string* active_log_ctx = &null_log_ctx;
+
+	struct log_ctx_setter {
+		log_ctx_setter(log_context& ctx) { celerity::detail::active_log_ctx = &ctx.value; }
+		~log_ctx_setter() { celerity::detail::active_log_ctx = &celerity::detail::null_log_ctx; }
+	};
+
+#define CELERITY_DETAIL_SET_LOG_CTX(ctx)                                                                                                                       \
```

Comment:
I feel like this macro makes code using it harder to understand, not easier. I wouldn't be opposed to name this guard type explicitly, like
```c++
detail::set_log_ctx push{ctx};
```
If we decide to keep the macro, I would advocate for adding a `__COUNTER__` to the variable name.

Reply:
I see your point in that it doesn't really convey it's RAII nature (which could be somewhat remedied by appending `_RAII` to the name). However I do think that mixing `SCREAM_CASE` macros with some `snake_case` type in `detail::` also doesn't make it easier to grok the connection between the two. The other alternative would be to move logging into variadic template functions that are defined as no-ops if the compile time log level doesn't match. Of course we'd then lose the option of logging source locations.

Adding the `__COUNTER__` would make sense, yes.

Reply:
I agree that a consistent interface should be used, but I also feel like making the RAII mechanism visible on the call-site is a good idea.

What about something like this:

```cpp
const auto raii_ctx = CELERITY_SET_LOG_CTX(ctx);
```

Reply:
@BlackMark29A that would stop us from completely removing the code at compile time.

I've been thinking, maybe it's enough to just improve the naming, e.g. to `CELERITY_DETAIL_LOG_CTX_GUARD`?

Reply:
> that would stop us from completely removing the code at compile time

Why is this a problem? It won't have any runtime overhead, if it's set to an empty struct, for example.

Reply:
Renamed to `CELERITY_LOG_SET_SCOPED_CTX`.